### ISSUE TITLE
Serving unbuilt directories -- module path transformation?

### ISSUE THREAD

**Issue by mercmobily** [Polymer/prpl-server#92]
Hi,

I am developing an app, and am using prpl-server as a module.
My `app.js` provides a bunch of custom URLs that are necessary to the app to function (as it should be).
I am starting to test the client side. Now... I am stuck -- help!
As far as I can tell, after _every_ modification I need to build the whole site, and reload it.
This is because if I want to just serve the unbuilt version of the site, I have two possible paths:

* If I serve the site with `polymer serve`, the module resolution happens on the fly  (`import { LitElement, html, css } from 'lit-element';` transforms that `lit-element` into a proper import path in node_modules) **BUT** I lose all of the site's routes that make my app actually work 

* If I serve the site with prpl-server (by "cheating" and naming a build `.`), I have all of my routes available, **BUT** there is no automatic import path resolution, which means that no library are loaded.

So, basically, without automatic import path resolution, the only chance to test every tiny change is by rebuilding, which can take up to 40 seconds.

Please tell me that:

* Either I missing something and it's possible to do what I want to do easily
* Or there is an easy, painless way to hack the automatic import path resolution into prpl-server

Thank you,

Merc.

**Comment by mercmobily** [comment]
I read [this blog post](https://www.polymer-project.org/blog/2018-02-26-3.0-preview-paths-and-names):

> We're adding support for bare specifiers to the Polymer CLI and related tools, as well as other projects like prpl-node-server.

Was this done? I am looking at the source code of prpl-server and can't find anything that will do automatic transformation of paths... maybe I am missing something?

**Comment by mercmobily** [comment]
Closing this. There is no point in this module doing it, while we can use polyserve in our express server. See https://github.com/Polymer/pwa-starter-kit/issues/358

Shame it's totally undocumented. Will do my best to get it documented.

**Comment by aomarks** [comment]
Hi @mercmobily. Sorry, the blog post isn't correct. We don't do the same kinds of automatic transformations in prpl-server that we do in polyserve, such as compilation to ES5, ES to AMD module transformation, or bare module specifier resolution.

The general idea is that for development you would use polyserve, where everything works automatically in every browser. For production, though, the idea is that you would first use polymer build to create a few static builds of your site, each targeted at some different set of browser capabilities, and then use prpl-server to differentially serve those static builds to the right browsers.

You may want to consider whether you really want the complexity of polyserve style automatic transformations running in your production server, because those transformations add latency (at least before they get cached), and also theoretically open your public service to a higher chance of hitting bugs or security issues. By doing the transformations up front with something like polymer build, your production server can be simpler and less prone to bugs.

**Comment by mercmobily** [comment]
We must have answered my post at the same time...
My problem (which is a real use-case problem) is that my application needs extra routes to work, and a simple polyserve won't provide them.
The solution (for posterity, or whoever stumbles upon this by accident) is to use polyserve as a module like I explain [here](https://github.com/Polymer/tools/issues/1077), specifically:

````
var startServer = require('polyserve/lib/start_server.js')

// .. various routes in my application, and then...

if (process.env.NODE_ENV === 'development' || !process.env.NODE_ENV){
  var polyserveApp = startServer.getApp({
    moduleResolution: 'node',
    npm: true,
  })
  app.use(polyserveApp)
} else {
  var prplHandler = prpl.makeHandler('build', polyConfigFile)
  app.get('/*', (req, res, next) => { prplHandler(req, res, next) })
}
````

I think this very snippet should be included in your documentation, as anybody who writes apps defining routes (surely the majority of devs!) would otherwise have to wait 30 seconds for their beast to build at every change.

Thanks!
